composer/loggers/progress_bar_logger.py

```diff
@@ -105,140 +128,178 @@ def __init__(
+            raise AssertionError('Cannot set pbar if self.is_train is not set.')
+
+        if self.is_train:
+            self.train_pbar = pbar
+        else:
+            self.eval_pbar = pbar
+
+    @property
+    def show_pbar(self) -> bool:
+        return self._show_pbar and dist.get_local_rank() == 0
+
     def log_data(self, state: State, log_level: LogLevel, data: Dict[str, Any]) -> None:
-        if dist.get_local_rank() == 0 and self.is_train in self.pbars:
+        # log to progress bar
+        if self.current_pbar:
             # Logging outside an epoch
-            assert self.is_train is not None
-            self.pbars[self.is_train].log_data(data)
-
-        if not self.should_log(state, log_level):
-            return
-        data_str = format_log_data_value(data)
-        log_str = f'[{log_level.name}][batch={int(state.timestamp.batch)}]: {data_str}'
-
-        if self.is_train in self.pbars:
-            # use tqdm.write to avoid interleaving with a progress bar
-            assert self.is_train is not None
-            self.pbars[self.is_train].pbar.write(log_str)
+            self.current_pbar.log_data(data)
+
+        # log to console
+        if self.should_log(state, log_level):
+            data_str = format_log_data_value(data)
+            log_str = f'[{log_level.name}][batch={int(state.timestamp.batch)}]: {data_str}'
+            self.log_to_console(log_str)
+
+    def log_to_console(self, log_str: str):
+        """Logs to the console, avoiding interleaving with a progress bar."""
+        if self.current_pbar:
+            # use tqdm.write to avoid interleaving
+            self.current_pbar.pbar.write(log_str)
         else:
             # write directly to self.stream; no active progress bar
             print(log_str, file=self.stream, flush=True)
 
-    def _start(self, state: State):
-        if dist.get_local_rank() != 0 or not self.show_pbar:
-            return
-        assert self.is_train is not None, 'self.is_train should be set by the callback'
-        assert state.dataloader_len is not None, 'dataloader_len should be set when using tqdm'
+    def _build_pbar(self, state: State, is_train: bool, epoch_style: bool = False) -> _ProgressBar:
+        """Builds a pbar that tracks in the units of max_duration.
+
+        Example:
+            Samples     train  73% ||███████████████        | 293873/400000
+
+        If epoch_style = True, then the pbar total will be the
+        numbers of batches in the epoch, regardless of the max_duration units.
+        This is often used to emit a pbar for each epoch, e.g.
+            Epoch     0 train 100%|█████████████████████████| 29/29
+            Epoch     1 train 100%|█████████████████████████| 29/29
+        """
+        position = 0 if is_train else 1
+        split = 'train' if is_train else 'val'
+
+        assert self.is_train is not None
+        if epoch_style:
+            assert state.dataloader_len is not None, 'dataloader_len should be set'
```

Comment:
Can we be graceful for this case? I think tqdm supports progress bars for an unknown total length

Reply:
done!